Yoast/wordpress-seo
# Fix social editor focus and implement better fallback/placeholder functionality

## Context


* The social form description focus was broken on free. This PR fixes that.
* The social preview did not always correctly show the fallbacks that were employed on the frontend, that is now fixed.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes an unreleased bug where social description input fields would not focus when clicking below the top sentence.
* [wordpress-seo-premium] Fixes a bug where social previews would not reflect the exact output on the frontend if a metadata description was set.

## Relevant technical choices:

* I have copied the state-managing code from the `social-metadata-previews` package into `wordpress-seo`. At some point, we should really abstract the forms to be state managed out of the box, because now each wrapper (the wrapper that combines the `upsell` with the `social form` in `free`, the wrapper that combines the `previews` with the `social form` in `premium`) has to implement the state management on it's own. Ideally, the form handles state management itself, and it can accepts a child that can also handle the state if it calls the right functions (e.g. the `Preview`). 
* Because that is a slightly bigger overhaul than fits this PR, I've opted to be pragmatic and just copy the code. It works. We hopefully don't have to adjust it too much. _(in the future I really want that refactor I mentioned above)_

## Test instructions

This PR can be tested by following these steps:

Javascript PR: https://github.com/Yoast/javascript/pull/729
Premium PR: https://github.com/Yoast/wordpress-seo-premium/pull/2842

* **TEST IN FREE**
* Checkout the javascript monorepo branch `use-preview-fallbacks-and-input-placeholders` and run `yarn link-monorepo`
* Verify that, when empty, both title and description input fields (social tab) have placeholders.
* Verify that focussing the description works by clicking anywhere in the description box.
* Note that tabbing is broken and that pressing `shift+up` on multiline texts does not select all texts. This is a `replacementVariableEditor` bug, outside of the scope of this PR.

* **TEST IN PREMIUM**
* In `premium`, checkout branch `use-preview-fallbacks-and-input-placeholders`, and run `git merge `free/use-preview-fallbacks-and-input-placeholders --no-commit`.
* Checkout the javascript monorepo branch `use-preview-fallbacks-and-input-placeholders` and run `yarn link-monorepo` in both `src` and `/premium`.
* Verify that everything works as above (focus, placeholder)
* Additionally, verify that if you set a meta description and title in the `snippet editor`, and leave the `social` inputs empty, the `social preview` description and title show the `snippet editor`'s meta description and title. This reflects the frontend better, as those values are also loaded onto `og:title` and `og:description`.

## UI changes

## Documentation

## Quality assurance


Fixes https://yoast.atlassian.net/browse/P1-22
